crates/tombi-schema-store/src/store.rs

```diff
@@ -1,18 +1,83 @@
-use std::{ops::Deref, sync::Arc};
-
+use crate::{
+    json::CatalogUrl, DocumentSchema, SchemaAccessor, SchemaAccessors, SchemaSpec, SchemaUrl,
+    SourceSchema,
+};
 use ahash::AHashMap;
-use futures::{future::BoxFuture, FutureExt};
-use itertools::Either;
+use async_trait::async_trait;
+use bytes::Bytes;
+use itertools::{Either, Itertools};
+use std::fmt::Debug;
+use std::{ops::Deref, sync::Arc};
 use tokio::sync::RwLock;
 use tombi_config::{Schema, SchemaOptions};
+use tombi_wasm_compat::box_future::{BoxFuture, Boxable};
+use tombi_wasm_compat::url::url_to_file_path;
 
-use crate::{
-    json::CatalogUrl, DocumentSchema, SchemaAccessor, SchemaAccessors, SchemaUrl, SourceSchema,
-};
+// todo: strip dependency to reqwest!
+
+#[cfg_attr(feature = "wasm", async_trait(?Send))]
+#[cfg_attr(not(feature = "wasm"), async_trait)]
+pub trait HttpClient: Debug + Send + Sync {
+    // fn new() -> Self;
+
+    async fn get_bytes(&self, url: &str) -> Result<Bytes, String>;
+}
+
+#[cfg(not(feature = "wasm"))]
+#[derive(Debug)]
+pub struct DefaultClient(reqwest::Client);
+
+#[cfg(not(feature = "wasm"))]
+impl DefaultClient {
+    pub fn new() -> Self {
+        Self(reqwest::Client::new())
+    }
+}
+
+#[cfg(not(feature = "wasm"))]
+#[async_trait]
+impl HttpClient for DefaultClient {
+    async fn get_bytes(&self, url: &str) -> Result<Bytes, String> {
+        let response = self
+            .0
+            .get(url)
+            .send()
+            .await
+            .map_err(|err| err.to_string())?;
+
+        let bytes = response.bytes().await.map_err(|err| err.to_string())?;
+        Ok(bytes)
+    }
+}
+
+#[cfg(feature = "wasm")]
+#[derive(Debug)]
+pub struct DefaultClient;
+
+#[cfg(feature = "wasm")]
+impl DefaultClient {
+    pub fn new() -> Self {
+        Self
+    }
+}
+
+#[cfg(feature = "wasm")]
+#[async_trait(?Send)]
+impl HttpClient for DefaultClient {
+    async fn get_bytes(&self, url: &str) -> Result<Bytes, String> {
+        let mut response = gloo_net::http::Request::get(url)
+            .send()
+            .await
+            .map_err(|err| err.to_string())?;
+
+        let bytes = response.binary().await.map_err(|err| err.to_string())?;
+        Ok(Bytes::from(bytes))
+    }
+}
 
 #[derive(Debug, Clone)]
 pub struct SchemaStore {
-    http_client: reqwest::Client,
+    http_client: Arc<dyn HttpClient>,
```

Comment:
Can we treat it statically as an impl HttpClient?

I haven't come up with a way to use it dynamically as it is determined at compile time.

Reply:
`SchemaStore` currently requires `Clone`. If we make `HttpClient` itself `Clone`, it will also require being `Sized`, making it impossible to use as a `dyn HttpClient`. To hold `HttpClient` as a clonable field without making the trait itself `Clone`, it seems necessary to wrap it in something like an `Arc`.

Alternatively, using generics is technically possible, but it would require propagating generics through `SchemaStore` and every dependent structure or function, which wouldn't be ideal.